### DISCUSSION TITLE
Bikeshedding names for `StoreTask.finish()`

### DISCUSSION THREAD

**Discussion by randomeizer** [pointfreeco/swift-composable-architecture#3116] [General]
It was mooted in [Slack](https://pointfreecommunity.slack.com/archives/C04KQQ7NXHV/p1716488902743429) that the `StoreTask.finish()` name feels ambiguous in its intention. The designed usage is something like:

```swift
await store.send(.action).finish()
```

However, because `finish` is an imperative verb, it may feel like it's a command to finish, rather than waiting for the task to finish.

There is some prior art from Apple for using `waitXXX`:

![cleanshot_2024-05-24_at_17 58 31_2x_720](https://github.com/pointfreeco/swift-composable-architecture/assets/3589315/f5f56acb-8539-4705-b33d-d2ab08e6a46d)

It does feel somewhat redundant having `await` and `wait` in the statement. For example:

```swift
await store.send(.action).waitUntilFinished()
```

Other possible options suggested include:

* `await store.send(.action).finished()`
* `await store.send(.action).untilFinished()`
* `await store.send(.action).hasFinished()`

None of these really have prior art in an async sense.

I would also say that it should make sense if just checking on the `StoreTask` directly. For example:

```swift
let myTask = await store.send(.action)

await myTask.waitUntilFinished() // clear, but redundant
await myTask.finished() // feels more like checking state than waiting for something
await myTask.untilFinished() // does not read like good English
await myTask.isFinished() // also feels like checking state, and `isCancelled` is not async.
await myTask.hasFinished() // reads ok, but no prior art
```

Of these, my favourite is actually probably `hasFinished` or `hasCompleted`. It's more clearly a 'past tense' phrase than `isFinished`. But it's not really out there much as a convention.

Otherwise, `waitUntilFinished` or `waitUntilCompleted` would be clearer than what we have at the moment.

What are people's thoughts or suggestions?

**Comment by jshier** [comment]
Use of `wait` is certainly redundant, and most of the examples in that list are old, synchronous interfaces. In fact, I'd say even `TaskGroup`'s APIs are misnamed, as `await group.waitForAll()` is rather ambiguous and redundant. Wait for all what? `await group.allFinished()` makes more sense to me. In that case, `.finished()` seems good enough for the singular version. 

I'd also like to point out that your dislike of various options due to checking state rather than waiting for something doesn't seem true, given the use of `await`. With `await`, the wait is implicit in all of these APIs. So really it's down to whether or not the API is grammatical and conveys the relevant information.

Personally, and given `finish()` can't function in a way that may be confused with it's current use (we can't command an arbitrary action to finish), and the use of `await` conveying the suspension point (wait over time), it's use seems fine to me. If the confusion is that it's something that happens in the future, and `await` isn't enough to convey that, then the past tense, `finished()` seem to convey that bit of information. If we consider `waitUntilFinished()` redundant, `untilFinished()` non grammatical, then `finished()` seems correct.

Unfortunately Apple has stopped updating the Swift naming guidelines as Swift adds new capabilities like native concurrency, so there's really no official guidance here.

**Reply by randomeizer** [reply]
Counter to that, there is `cancel()` and `isCancelled`. `finished()` as a function is somewhere between the two. Not sure that's ideal.

An `isFinished` property would probably not be appropriate, although you have examples like `Task.value` which is async. But the fact that `isFinished` would be waiting until the task is finished makes it different to `isCancelled`, which is a synchronous call.

**Comment by randomeizer** [comment]
Another possibility that occurred. `Task` has `value` and `result` which allows access to the final outcome of the task, waiting for it to conclude. `StoreTask` doesn't have a value type, and can never fail, but it does have two possible outcomes: finished and cancelled.

What if we had an enum that indicated that outcome, and a property to check the result. Something like:

```swift
extension StoreTask {
  public enum Completion {
    case finished
    case cancelled
  }
  
  public var completion: Completion {
    get async {
      await self.rawValue?.cancellableValue
      return self.isCancelled ? .cancelled : .finished
     }
  }
}
```

This would then be used as:

```swift
await store.send(.action).completion
```

**Reply by jshier** [reply]
All tasks finish, but not all are cancelled. This distinction doesn't seem to help, and really makes things more complicated than necessary.

**Reply by jshier** [reply]
I'd also point out that `completion` and `finish` are synonymous, so you're not really conveying anything more than the API already does.

**Reply by randomeizer** [reply]
Yeah, `finished` might not be the right word here. The intention is that it flags it got to the end of the task and was not canceled. Open to suggestions.

However, I don't believe `completion` and `finish` are synonyms exactly. For one, `completion` is only a noun, meaning `the state of being finished`, whereas `finish` is both a verb and a noun, and more often used as a verb, which is partly what is causing confusion here.

As for not adding to the API, I'd say that this version actually makes it clear if the task reached the end or was cancelled, and the current API does not, unless you check for the negative of `isCancelled`.

**Reply by jshier** [reply]
There's no intention here to distinguish between finishing while cancelled or not canceled. It's simply to indicate the end of the action. We could say `.done()`, perhaps that's more clear, but I don't think it's enough of an improvement to justify a change, nor does it align with other common APIs.

**Reply by randomeizer** [reply]
Why not allow distinction though? Task lets you distinguish between success and failure via `result`. We could call it `result` too, but I think that would imply an actual `Result`, which is not the case here.

There are also `Subscribers.Completion` from Combine, used by `Sink`, etc., which is defined as:

```swift
    /// A signal that a publisher doesn’t produce additional elements, either due to normal completion or an error.
    @frozen public enum Completion<Failure> where Failure : Error {

        /// The publisher finished normally.
        case finished

        /// The publisher stopped publishing due to the indicated error.
        case failure(Failure)
    }
```

If the point is that cancellations aren't usually communicated so directly, sure. Is that a good thing though?

If it's the phrase that's the issue, here's some possible other terms for "finished and didn't get cancelled":

* resolved
* terminated
* ended
* finalised
* closed

**Reply by jshier** [reply]
If you do want to know how the task finished (not sure why, but it may be useful), and you don't want to use `finished() -> Completion`, perhaps `completed() -> Completion` would be appropriate, as long as you could ignore the value. `await store.send(.action).completed()` But the return value doesn't really give you anything that simply checking an `isCancelled` property after the fact wouldn't. In fact, if `completed()` simply returned the task itself, I think you could do `let isCancelled = await store.send(.action).completed().isCancelled`, if having it on one line was important.

**Reply by randomeizer** [reply]
Sure, those are possibilities. Grammatically speaking, `await ... completion` is more correct than `await ... completed`, but there are a lot of other grammatically incorrect combinations with `await` all over the place, so not that big a deal.

Ignoring `completion` as a computed variable is possible, but requires `_ = store.send(...).completion`. Making it a function would allow using `@discardableResult`, but would be less of a match with `Task.value` and `Task.result`.

**Reply by jshier** [reply]
You must consider the whole usage to determine whether it reads grammatically, you can't just cut out the middle. `await store.send(.action).completed()` seems just as grammatical (wait until the store's send is completed) as `await store.send(.action).completion` (wait for the completion value of the store's send). 

And given that you'll usually want to ignore the `completion` value, unlike `Task.value`, it seems like we'd want an interface you can actually ignore.

**Reply by randomeizer** [reply]
And here we're back to adding words back into the prefix, such as `untilCompletion`, `hasCompleted`, or something along those lines.

**Comment by jegnux** [comment]
I like completion but I don't think a car can be @discardableResult so it would require to write 
```swift
_ = await store.send(.action).completion
```

What about having an async counterpart to .send() that just conveys the fact that it does not just send but also wait for all effects to run?

```swift
await store.run(.action)
```

Or eventually 
```swift
await store.reduce(.action)
```

**Reply by randomeizer** [reply]
Correct, computed `var` values can't be discarded, (although this does seem to compile, which is interesting):

```swift
var foo { 
  @discardableResult
  get async { ... }
}
```

An async alternate to `send` could work, and potentially be clearer. I think `run` is not necessarily clear as far as expecting it to wait. Maybe `sendAndFinish(.action)` or something? This is getting back to the deprecated API though.

**Reply by Mackarous** [reply]
I think `run` is clear considering we already use it within the reducer pattern

**Reply by randomeizer** [reply]
True that.
